doc/developer/design/20220203_connectors.md

```diff
@@ -0,0 +1,211 @@
+SINK` make this a verbose affair. The problem is particularly acute with
+Avro-formatted Kafka sources that configure authentication:
+
+```sql
+CREATE SOURCE kafka1
+FROM KAFKA BROKER 'kafka:9092' TOPIC 'top1' WITH (
+    security_protocol = 'SASL_SSL',
+    sasl_mechanisms = 'PLAIN',
+    sasl_username = 'username',
+    sasl_password = 'password',
+)
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY 'https://schema-registry' WITH (
+    username = 'username',
+    password = 'password'
+);
+
+CREATE SOURCE kafka2
+FROM KAFKA BROKER 'kafka:9092' TOPIC 'top2' WITH (
+    security_protocol = 'SASL_SSL',
+    sasl_mechanisms = 'PLAIN',
+    sasl_username = 'username',
+    sasl_password = 'password'
+)
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY 'https://schema-registry' WITH (
+    username = 'username',
+    password = 'password'
+);
+```
+
+These two source definition differ only in their topic specification, but
+must duplicate eight other configuration parameters.
+
+With the connectors proposed in this document, the `CREATE SOURCE` can instead
+share all the relevant configuration:
+
+```sql
+CREATE CONNECTOR kafka FOR
+KAFKA BROKER 'kafka:9092' WITH (
+    security_protocol = 'SASL_SSL',
+    sasl_mechanisms = 'PLAIN',
+    sasl_username = 'username',
+    sasl_password = 'password'
+);
+
+CREATE CONNECTOR schema_registry FOR
+CONFLUENT SCHEMA REGISTRY 'https://schema-registry' WITH (
+    username = 'username',
+    password = 'password'
+);
+
+CREATE SOURCE kafka1
+FROM KAFKA CONNECTOR kafka TOPIC 'top1'
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTOR schema_registry;
+
+CREATE SOURCE kafka2
+FROM KAFKA CONNECTOR kafka TOPIC 'top2'
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTOR schema_registry;
+```
+
+## Design
+
+### SQL syntax and semantics
+
+Connector create syntax will come in type specific variants and will not include `with_options`, instead completely specifying the potential options in each variant. 
+
+Using connectors syntactically means replacing the source specific keyword such as BROKER or CONNECTION with CONNECTOR and supplying one or more connector names. 
+
+Properties will be merged by applying values in the supplied order with the last write winning, this supports a usage pattern of supplying base values and then progressively more specific overrides while remaining easy to reason about.  
+
+As long as connectors do not support ALTER then it is strictly not necessary to store the relationship between source and connector in the catalog, instead during planning of the CREATE SOURCE statement the values from the connector can be substituted into relevant places in the `CreateSourceConnector` without requiring any parsing changes for any part of the `CreateSourceStatement`. 
+
+While technically supporting ALTER does not have to require changing the catalog representation of sources, the utility of supporting ALTER without CASCADE is very small and supporting CASCADE requires that relations are discoverable.
```

Comment:
This is fine if we don't care about changing the state of "instantiated sources", the `CASCADE` seems correct here if we want the change to propagate out to those leaf nodes as had been discussed in slack.

Reply:
> This is fine if we don't care about changing the state of "instantiated sources", the `CASCADE` seems correct here if we want the change to propagate out to those leaf nodes as had been discussed in slack.

Here is my model of the situation. Let's suppose we have a `kafka_connectors` table and a `kafka_sources` table with the following schemas:

```sql
CREATE TABLE kafka_connectors (
    id int,
    name text,
    broker text
)

CREATE TABLE kafka_sources (
    id int,
    name text,
    topic text,
    kafka_connector_id int NOT NULL REFERENCES kafka_connectors (id) ON DELETE CASCADE
)
```

In this model, an `ALTER CONNECTOR` maps to an `UPDATE kafka_connectors` query. Nothing cascades during an `UPDATE` of `kafka_connectors`, because each property lives in precisely one place. By updating the connector property in `kafka_connectors`, the next person to join `kafka_connectors` and `kafka_sources` will see the updated value; no cascade required.

`DROP CONNECTOR` maps to `DELETE FROM kafka_connectors`, where the `RESTRICT` vs `CASCADE` behavior comes into play. The reason that you need to talk about `RESTRICT`/`CASCADE` for is because you would otherwise violate referential integrity if you removed a connector from `kafka_connectors` that was still referenced by `kafka_sources`. 

I suppose you could look at this as "`ALTER`s always `CASCADE`", but the point is that you never explicitly attach `CASCADE` to `ALTER`, nor is there such a thing as `ALTER ... RESTRICT`!

Reply:
Thank you, this really helped things click into place. If I am understanding correctly then conceptually we would be changing `ExternalConnector::Kafka` from always being an embedded `KafkaSourceConnector` to being either embedded or a reference, and during planning whenever its a reference we would simply do the lookup in the catalog to populate an instance of it?

Reply:
Yes exactly!